src/query/parser/promql/parse.go

```diff
@@ -148,6 +153,21 @@ func (p *parseState) addLazyOffsetTransform(offset time.Duration) error {
 	return nil
 }
 
+func adjustOffset(offset time.Duration, step time.Duration) time.Duration {
+	if offset == 0 {
+		return offset
+	}
+
+	align := offset % step
+	if align != 0 {
+		// NB: Prometheus rounds offsets up to step size, e.g. a 61 second offset with
+		// a 1 minute stepsize gets rounded to a 2 minute offset.
+		align -= step
+	}
+
+	return offset - align
```

Comment:
Hm interesting, seems like this would read more straight forward if it was simply:

```go


func adjustOffset(offset time.Duration, step time.Duration) time.Duration {
	if offset == 0 {
		return offset
	}
	if offset%step == 0 {
		return offset
	}
	// Perform integer division to round down to nearest step, then add step to match Prometheus rounding up
	return (offset / step) + step
}
```